src/order/cover.lean

```diff
@@ -9,17 +9,108 @@ import data.set.intervals.ord_connected
 # The covering relation
 
 This file defines the covering relation in an order. `b` is said to cover `a` if `a < b` and there
-is no element in between. ∃ b, a < b
+is no element in between. We say that `b` weakly covers `a` if `b` covers `a` or `b = a`.
 
 ## Notation
 
-`a ⋖ b` means that `b` covers `a`.
+* `a ⋖ b` means that `b` covers `a`.
+* `a ⩿ b` means that `b` covers `a` or `b = a`.
 -/
 
-open set
+open set order_dual
 
 variables {α β : Type*}
 
+section weakly_covers
+
+section preorder
+variables [preorder α] [preorder β] {a b c: α} {f : α ↪o β} {e : α ≃o β}
+
+/-- `wcovby a b` means that `a = b` or `b` covers `a`.
+This means that `a ≤ b` and there is no element in between.
+-/
+def wcovby (a b : α) : Prop := a ≤ b ∧ ∀ ⦃c⦄, a < c → ¬ c < b
+
+infix ` ⩿ `:50 := wcovby
+
+lemma wcovby.le (h : a ⩿ b) : a ≤ b := h.1
+
+lemma wcovby.refl (a : α) : a ⩿ a := ⟨le_rfl, λ c hc, hc.not_lt⟩
+lemma wcovby.rfl : a ⩿ a := wcovby.refl a
+
+lemma wcovby_of_eq_or_eq (hab : a ≤ b) (h : ∀ c, a ≤ c → c ≤ b → c = a ∨ c = b) : a ⩿ b :=
+begin
+  refine ⟨hab, λ c hac hcb, _⟩,
+  rcases h c hac.le hcb.le with rfl|rfl; exact lt_irrefl c ‹c < c›
+end
+
+/-- If `a ≤ b`, then `b` does not cover `a` iff there's an element in between. -/
+lemma not_wcovby_iff (h : a ≤ b) : ¬ a ⩿ b ↔ ∃ c, a < c ∧ c < b :=
+by simp_rw [wcovby, h, true_and, not_forall, exists_prop, not_not]
+
+instance wcovby.is_refl : is_refl α (⩿) := ⟨λ _, wcovby.rfl⟩
+
+lemma wcovby.Ioo_eq (h : a ⩿ b) : Ioo a b = ∅ :=
+eq_empty_iff_forall_not_mem.2 $ λ x hx, h.2 hx.1 hx.2
+
+lemma wcovby.of_image (h : f a ⩿ f b) : a ⩿ b :=
+⟨f.le_iff_le.mp h.le, λ c hac hcb, h.2 (f.lt_iff_lt.mpr hac) (f.lt_iff_lt.mpr hcb)⟩
+
+lemma wcovby.image (hab : a ⩿ b) (h : (range f).ord_connected) : f a ⩿ f b :=
+begin
+  refine ⟨f.monotone hab.le, λ c ha hb, _⟩,
+  obtain ⟨c, rfl⟩ := h.out (mem_range_self _) (mem_range_self _) ⟨ha.le, hb.le⟩,
+  rw f.lt_iff_lt at ha hb,
+  exact hab.2 ha hb,
+end
+
+protected lemma set.ord_connected.apply_wcovby_apply_iff (h : (range f).ord_connected) :
+  f a ⩿ f b ↔ a ⩿ b :=
+⟨λ h2, h2.of_image, λ hab, hab.image h⟩
+
+@[simp] lemma order_iso.apply_wcovby_apply_iff (e : α ≃o β) : e a ⩿ e b ↔ a ⩿ b :=
```

Comment:
Can this be stated in terms of `order_iso_class`?

Reply:
That requires a `has_le` instance, and I don't think it is a good idea to add such an instance.

Reply:
What do you mean? What I'm saying is the following
```suggestion
@[simp] lemma apply_wcovby_apply_iff {E : Type*} [order_iso_class E α β] (e : E) :
  e a ⩿ e b ↔ a ⩿ b :=
```

Reply:
Oh I see... I thought you meant a new instance of the `order_iso_class`.

Reply:
I tried, but I don't know if I can.

I generalized the `ord_connected` lemmas according to this way of formulating (keeping also the `order_iso` lemmas for projection notation - is that standard?)
But for the proof I'm using `order_embedding`, and I don't think there is an `order_embedding_class` that I can use.

Reply:
I don't understand. Nowhere does your code use `order_embedding`, nor is anything in the `order_embedding` API not accessible through the `order_iso_class` one. Doesn't `(ord_connected_range e).apply_wcovby_apply_iff` work?

Also, we do not keep the `order_iso` lemmas around for dot notation.

Reply:
`ord_connected.apply_wcovby_apply_iff` has an `order_embedding` argument... Also, `order_iso_class` extends `equiv_like`, so I don't know how it would apply to `order_iso_class`.

Reply:
Ah sorry, I missed it. You can then use `(e : α ≃o β).to_order_embedding`.

Reply:
Ok